Source/Core/Common/x64Emitter.h

```diff
@@ -1090,29 +1005,148 @@ class XEmitter
   // Utility functions
   // The difference between this and CALL is that this aligns the stack
   // where appropriate.
-  void ABI_CallFunction(const void* func);
-
-  void ABI_CallFunctionC16(const void* func, u16 param1);
-  void ABI_CallFunctionCC16(const void* func, u32 param1, u16 param2);
-
-  // These only support u32 parameters, but that's enough for a lot of uses.
-  // These will destroy the 1 or 2 first "parameter regs".
-  void ABI_CallFunctionC(const void* func, u32 param1);
-  void ABI_CallFunctionCC(const void* func, u32 param1, u32 param2);
-  void ABI_CallFunctionCP(const void* func, u32 param1, void* param2);
-  void ABI_CallFunctionCCC(const void* func, u32 param1, u32 param2, u32 param3);
-  void ABI_CallFunctionCCP(const void* func, u32 param1, u32 param2, void* param3);
-  void ABI_CallFunctionCCCP(const void* func, u32 param1, u32 param2, u32 param3, void* param4);
-  void ABI_CallFunctionPC(const void* func, void* param1, u32 param2);
-  void ABI_CallFunctionPPC(const void* func, void* param1, void* param2, u32 param3);
-  void ABI_CallFunctionAC(int bits, const void* func, const OpArg& arg1, u32 param2);
-  void ABI_CallFunctionA(int bits, const void* func, const OpArg& arg1);
+  template <typename FunctionPointer>
+  void ABI_CallFunction(FunctionPointer func)
+  {
+    static_assert(std::is_pointer<FunctionPointer>() &&
+                      std::is_function<std::remove_pointer_t<FunctionPointer>>(),
+                  "Supplied type must be a function pointer.");
+
+    const void* ptr = reinterpret_cast<const void*>(func);
+    const u64 address = reinterpret_cast<u64>(ptr);
+    const u64 distance = address - (reinterpret_cast<u64>(code) + 5);
+
+    if (distance >= 0x0000000080000000ULL && distance < 0xFFFFFFFF80000000ULL)
+    {
+      // Far call
+      MOV(64, R(RAX), Imm64(address));
+      CALLptr(R(RAX));
+    }
+    else
+    {
+      CALL(ptr);
+    }
+  }
+
+  template <typename FunctionPointer>
+  void ABI_CallFunctionC16(FunctionPointer func, u16 param1)
```

Comment:
What do you think about a generic
template<typename T> OpArg TooImm(T param)
function to hide all kind of different parameters? We likely need it for u8, u16, u32, u64, and void*.